File: scratchpad/w-equiv-actorjoin28.cpp
#include "runtime/protocol/service_wire_codec.hpp"
#include <service_wire_pilot_codec.hpp>
#include <iostream>
#include <iomanip>
#include <sstream>

using namespace zlink::framework::runtime;

static std::string hex(const std::vector<std::uint8_t>& v) {
    std::ostringstream o;
    for (auto b : v) o << std::hex << std::setfill('0') << std::setw(2) << (int)b;
    return o.str();
}

int main() {
    protocol::actor_route_fence_t actor{"actor", 2, {0x01,0x02}, 3, 4, 5};
    protocol::spot_route_fence_t spot{"spot", 6, {0x07,0x08}, 9, 10, 11};
    protocol::actor_join_request_t req{1, actor, true, spot};
    auto handBytes = protocol::encode_actor_join_request(req);

    protocol::service_wire_pilot_fence gActor{"actor", 2, {0x01,0x02}, 3, 4, 5};
    protocol::service_wire_pilot_fence gSpot{"spot", 6, {0x07,0x08}, 9, 10, 11};
    protocol::service_wire_pilot_actor_join_28 greq{1, gActor, true, gSpot};
    auto genBytes = protocol::encode_actor_join_28(greq);

    std::cout << "hand:      " << hex(handBytes) << "\n";
    std::cout << "generated: " << hex(genBytes) << "\n";
    std::cout << "equal: " << (handBytes == genBytes ? "true" : "false") << "\n";
    return handBytes == genBytes ? 0 : 1;
}

File: scratchpad/WEquivActorJoin28.java
import systems.zlink.framework.runtime.protocol.ServiceWirePilotCodec;
import java.util.HexFormat;

public class WEquivActorJoin28 {
    public static void main(String[] args) throws Exception {
        var actor = new ServiceWirePilotCodec.Fence("actor", 2L, new byte[]{1,2}, 3L, 4L, 5L);
        var spot = new ServiceWirePilotCodec.Fence("spot", 6L, new byte[]{7,8}, 9L, 10L, 11L);
        var join = new ServiceWirePilotCodec.ActorJoin28(1L, actor, true, spot);
        byte[] bytes = ServiceWirePilotCodec.encodeActorJoin28(join);
        String hex = HexFormat.of().formatHex(bytes);
        String expected = "5a4d011c000000000000000001056163746f720000000000000002020102000000000000000300000000000000040000000000000005010473706f7400000000000000060207080000000000000009000000000000000a000000000000000b";
        System.out.println("generated: " + hex);
        System.out.println("expected : " + expected);
        System.out.println("equal: " + hex.equals(expected));
    }
}
